File: backend/app/services/github_service.py
import re

from github import Github, Auth

from app.core.config import get_settings
from app.services.vcs_provider import FileDiff, PRDiff


class GitHubService:
    """GitHub VCS provider using PyGithub."""

    def __init__(self) -> None:
        self._token = get_settings().github_token

    def _get_client(self) -> Github:
        return Github(auth=Auth.Token(self._token))

    def fetch_pr_diff(self, pr_url: str) -> PRDiff:
        """Fetch structured diff from a GitHub PR URL."""
        match = re.match(r"https?://github\.com/([^/]+/[^/]+)/pull/(\d+)", pr_url)
        if not match:
            raise ValueError(f"Invalid PR URL: {pr_url}")

        repo_name = match.group(1)
        pr_number = int(match.group(2))

        gh = self._get_client()
        repo = gh.get_repo(repo_name)
        pr = repo.get_pull(pr_number)

        files = []
        total_add = 0
        total_del = 0

        for f in pr.get_files():
            files.append(
                FileDiff(
                    filename=f.filename,
                    patch=f.patch or "",
                    additions=f.additions,
                    deletions=f.deletions,
                    status=f.status,
                )
            )
            total_add += f.additions
            total_del += f.deletions

        return PRDiff(
            repo_name=repo_name,
            pr_number=pr_number,
            title=pr.title,
            body=pr.body or "",
            base_branch=pr.base.ref,
            head_branch=pr.head.ref,
            files=files,
            total_additions=total_add,
            total_deletions=total_del,
        )

    def post_inline_comments(self, pr_url: str, findings: list[dict]) -> int:
        """Post review findings as inline comments on a GitHub PR."""
        match = re.match(r"https?://github\.com/([^/]+/[^/]+)/pull/(\d+)", pr_url)
        if not match:
            raise ValueError(f"Invalid PR URL: {pr_url}")

        repo_name = match.group(1)
        pr_number = int(match.group(2))

        gh = self._get_client()
        repo = gh.get_repo(repo_name)
        pr = repo.get_pull(pr_number)

        # Build review comments
        comments = []
        for finding in findings:
            body = f"**[{finding['severity'].upper()}]** {finding['title']}\n\n{finding['description']}"
            if finding.get("suggested_fix"):
                body += f"\n\n```suggestion\n{finding['suggested_fix']}\n```"
            body += f"\n\n_Agent: {finding['agent_role']} | Confidence: {finding['confidence']:.0%}_"

            comment = {
                "path": finding["file_path"],
                "body": body,
            }
            if finding.get("line_number"):
                comment["line"] = finding["line_number"]
                comment["side"] = "RIGHT"

            comments.append(comment)

        if not comments:
            return 0

        # Post as a single review with all comments
        pr.create_review(
            body=f"## SwarmForge Code Review\n\nFound **{len(comments)}** issues across {len(set(c['path'] for c in comments))} files.",
            event="COMMENT",
            comments=comments,
        )

        return len(comments)

File: desktop/server/github-service.ts
import { Octokit } from "@octokit/rest";

export interface FileDiff {
  filename: string;
  patch: string;
  additions: number;
  deletions: number;
  status: string;
}

export interface PRDiff {
  repoName: string;
  prNumber: number;
  title: string;
  body: string;
  baseBranch: string;
  headBranch: string;
  files: FileDiff[];
  totalAdditions: number;
  totalDeletions: number;
}

export interface ParsedPRUrl {
  owner: string;
  repo: string;
  number: number;
  platform: string;
}

const SKIP_PATTERNS = [
  ".lock",
  ".sum",
  "package-lock.json",
  "yarn.lock",
  ".min.js",
  ".min.css",
  ".map",
  ".png",
  ".jpg",
  ".svg",
  ".ico",
  ".woff",
  ".woff2",
];

function shouldSkip(filename: string): boolean {
  return SKIP_PATTERNS.some((ext) => filename.endsWith(ext));
}

export function parsePRUrl(url: string): ParsedPRUrl | null {
  // GitHub
  const ghMatch = url.match(
    /https?:\/\/github\.com\/([^/]+)\/([^/]+)\/pull\/(\d+)/
  );
  if (ghMatch) {
    return {
      owner: ghMatch[1],
      repo: ghMatch[2],
      number: parseInt(ghMatch[3], 10),
      platform: "github",
    };
  }

  // GitLab (basic — for display only, not API)
  const glMatch = url.match(
    /https?:\/\/[^/]+\/(.+?)\/-\/merge_requests\/(\d+)/
  );
  if (glMatch) {
    return {
      owner: "",
      repo: glMatch[1],
      number: parseInt(glMatch[2], 10),
      platform: "gitlab",
    };
  }

  return null;
}

export async function fetchPRDiff(
  prUrl: string,
  token: string
): Promise<PRDiff> {
  const parsed = parsePRUrl(prUrl);
  if (!parsed || parsed.platform !== "github") {
    throw new Error("Only GitHub PRs supported in desktop mode");
  }

  const octokit = new Octokit({ auth: token });

  const { data: pr } = await octokit.pulls.get({
    owner: parsed.owner,
    repo: parsed.repo,
    pull_number: parsed.number,
  });

  const { data: prFiles } = await octokit.pulls.listFiles({
    owner: parsed.owner,
    repo: parsed.repo,
    pull_number: parsed.number,
    per_page: 100,
  });

  const files: FileDiff[] = [];
  let totalAdditions = 0;
  let totalDeletions = 0;

  for (const f of prFiles) {
    if (shouldSkip(f.filename) || !f.patch) continue;
    files.push({
      filename: f.filename,
      patch: f.patch,
      additions: f.additions,
      deletions: f.deletions,
      status: f.status,
    });
    totalAdditions += f.additions;
    totalDeletions += f.deletions;
  }

  return {
    repoName: `${parsed.owner}/${parsed.repo}`,
    prNumber: parsed.number,
    title: pr.title,
    body: pr.body || "",
    baseBranch: pr.base.ref,
    headBranch: pr.head.ref,
    files,
    totalAdditions,
    totalDeletions,
  };
}

export interface Finding {
  severity: string;
  file_path: string;
  line_number?: number | null;
  title: string;
  description: string;
  suggested_fix?: string | null;
  agent_role?: string;
  confidence?: number;
}

export async function postInlineComments(
  prUrl: string,
  token: string,
  findings: Finding[]
): Promise<number> {
  const parsed = parsePRUrl(prUrl);
  if (!parsed || parsed.platform !== "github") return 0;

  const octokit = new Octokit({ auth: token });

  const comments = findings
    .filter((f) => f.file_path)
    .map((f) => {
      let body = `**[${f.severity.toUpperCase()}]** ${f.title}\n\n${f.description}`;
      if (f.suggested_fix) {
        body += `\n\n\`\`\`suggestion\n${f.suggested_fix}\n\`\`\``;
      }
      body += `\n\n_Agent: ${f.agent_role ?? "unknown"} | Confidence: ${Math.round((f.confidence ?? 0.5) * 100)}%_`;

      const comment: Record<string, unknown> = {
        path: f.file_path,
        body,
      };
      if (f.line_number) {
        comment.line = f.line_number;
        comment.side = "RIGHT";
      }
      return comment;
    });

  if (!comments.length) return 0;

  const uniqueFiles = new Set(comments.map((c) => c.path as string)).size;

  type ReviewComment = NonNullable<
    Parameters<typeof octokit.pulls.createReview>[0]["comments"]
  >[number];

  await octokit.pulls.createReview({
    owner: parsed.owner,
    repo: parsed.repo,
    pull_number: parsed.number,
    body: `## SwarmForge Code Review\n\nFound **${comments.length}** issues across ${uniqueFiles} files.`,
    event: "COMMENT",
    comments: comments as ReviewComment[],
  });

  return comments.length;
}
